plugins/mongodb/mongodbembed.cpp

```diff
@@ -0,0 +1,2302 @@
+ * Here is some example code of the plugin being used.
+ */
+namespace mongodbembed
+{
+    const time_t OBJECT_EXPIRE_TIMEOUT_SECONDS = 60 * 2; // Two minutes
+    static std::once_flag CONNECTION_CACHE_INIT_FLAG;
+
+    //--------------------------------------------------------------------------
+    // Plugin Classes
+    //--------------------------------------------------------------------------
+
+    /**
+     * @brief Takes an exception object and outputs the error to the user
+     * 
+     * @param e mongocxx::exception has some values that are useful to look for
+     */
+    void reportIfQueryFailure(const mongocxx::exception &e)
+    {
+        if (e.code() == mongocxx::error_code::k_invalid_collection_object) 
+        {
+            failx("Collection not found: %s",e.what());
+        }
+        if (e.code().value() == 26) 
+        {
+            failx("NamespaceNotFound: %s",e.what());
+        }
+        if (e.code().value() == 11000) 
+        {
+            failx("Duplicate Key: %s",e.what());
+        }
+        failx("Exception: %s", e.what());
+    }
+
+    /**
+     * @brief deserialize MongoDB Extended JSON relaxed format into normal JSON for the RowStream class
+     * Examples of EJSON relaxed format are:
+     * "dateField": {"$date":{"$numberLong":"1565546054692"}}
+     * "doubleField": {"$numberDouble":"10.5"}
+     * For more documentation on EJSON. https://www.mongodb.com/docs/manual/reference/mongodb-extended-json/
+     * 
+     * @param result Reference to a result string where the standard JSON should be written.
+     * @param row Pointer to the beginning of the result row.
+     */
+    void deserializeEJSON(std::string &result, const char * row)
+    {
+        auto start = row;
+        int depth = 0; // Keeps track of depth so we don't exit on wrong '}'
+        const char * lastBrkt;
+
+        while (*start != '}' || depth > 1) 
+        {
+            if (*start == '{') // If we see a document increase the depth
+            {
+                depth++;
+                lastBrkt = start; // Keep track of last bracket in case we need to backtrack
+            }
+            if (*start == '$' && *(start-1) == '\"') // Look for '$' to mark the start of a datatype
+            { 
+                auto end = start;
+                while (*end != '\"') end++;
+                auto key = std::string(start, end - start); // Get datatype
+                result += std::string(row, lastBrkt - row); // Add everything before we went into nested document
+                if (key == "$regularExpression") // Some data types are unsupported as they are not straightforward to deserialize
+                {
+                    UNSUPPORTED("Regular Expressions are currently Unsupported"); // TO DO handle unsupported types by not throwing an exception.
+                } 
+                else if (key == "$timestamp") 
+                {
+                    while (*end != '}') end++; // Skip over timestamp
```

Comment:
are there other chars which if encountered, should force a break out of the loop? should we ever iterate past string terminator, or len of string-1?

Reply:
For this case we are looking at an object of this form:`{"someTimestamp": {"$timestamp":{"t":1565545664,"i":1}}`. Since it is not trivial to convert this to an ECL data type I simply skip over it and replace it with an empty string. I believe for this case we will only want to break the loop when a '}' is encountered and we need to do that twice for timestamps. There shouldn't be a reason where we need to iterate past the string terminator or len of string-1.